File: js/core/ClaimLedger.js
// ============================================================
// NexusResearch -- ClaimLedger
// Verified claims repository with confidence scoring
// ============================================================

import { uid } from '../utils/helpers.js';

export class ClaimLedger {
  constructor() {
    this._claims = new Map();
    this._listeners = new Set();
  }

  /**
   * Add a new claim.
   * @param {Object} claimDef
   * @param {string} claimDef.text         - The claim statement
   * @param {number} claimDef.confidence   - 0-100
   * @param {string} claimDef.sourceAgent  - Agent that produced this claim
   * @param {string} [claimDef.topic]      - Related topic/sub-task
   * @param {string[]} [claimDef.evidence] - Evidence descriptions
   * @returns {Object} The created claim
   */
  addClaim(claimDef) {
    const claim = {
      id: uid(),
      text: claimDef.text,
      confidence: claimDef.confidence || 50,
      sourceAgent: claimDef.sourceAgent,
      contributingAgents: [claimDef.sourceAgent],
      topic: claimDef.topic || '',
      evidence: claimDef.evidence || [],
      debateId: null,
      status: 'unverified',   // unverified | disputed | verified | rejected
      createdAt: Date.now(),
    };
    this._claims.set(claim.id, claim);
    this._emit('add', claim);
    return claim;
  }

  /**
   * Update confidence score.
   */
  updateConfidence(claimId, newConfidence, agentId = null) {
    const claim = this._claims.get(claimId);
    if (!claim) return;
    claim.confidence = Math.max(0, Math.min(100, newConfidence));
    if (agentId && !claim.contributingAgents.includes(agentId)) {
      claim.contributingAgents.push(agentId);
    }
    this._emit('update', claim);
  }

  /**
   * Mark a claim's status.
   */
  setStatus(claimId, status, debateId = null) {
    const claim = this._claims.get(claimId);
    if (!claim) return;
    claim.status = status;
    if (debateId) claim.debateId = debateId;
    this._emit('update', claim);
  }

  /**
   * Add evidence to a claim.
   */
  addEvidence(claimId, evidence, agentId) {
    const claim = this._claims.get(claimId);
    if (!claim) return;
    claim.evidence.push(evidence);
    if (agentId && !claim.contributingAgents.includes(agentId)) {
      claim.contributingAgents.push(agentId);
    }
    this._emit('update', claim);
  }

  /** Get all claims, optionally filtered by min confidence */
  getClaims(minConfidence = 0) {
    return [...this._claims.values()].filter(c => c.confidence >= minConfidence);
  }

  /** Get claims by status */
  getByStatus(status) {
    return [...this._claims.values()].filter(c => c.status === status);
  }

  /** Get a single claim */
  get(claimId) {
    return this._claims.get(claimId) || null;
  }

  /** Get all claims sorted by confidence descending */
  getSorted() {
    return [...this._claims.values()].sort((a, b) => b.confidence - a.confidence);
  }

  /** Average confidence across all claims */
  get averageConfidence() {
    const claims = [...this._claims.values()];
    if (claims.length === 0) return 0;
    return Math.round(claims.reduce((sum, c) => sum + c.confidence, 0) / claims.length);
  }

  get size() { return this._claims.size; }

  /** Clear all claims */
  clear() { this._claims.clear(); }

  /** Subscribe to claim events */
  on(callback) {
    this._listeners.add(callback);
    return () => this._listeners.delete(callback);
  }

  _emit(event, claim) {
    for (const cb of this._listeners) {
      try { cb(event, claim); } catch (e) { console.error('[ClaimLedger]', e); }
    }
  }
}

File: nexus/core/claim_ledger.py
"""
Claim Ledger Module
Tracks claims, evidence chains, contributing agents, and confidence scores.
"""

import asyncio
import time
import uuid
from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class Claim(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    text: str
    confidence: int = 50
    source_agent: str
    contributing_agents: List[str] = Field(default_factory=list)
    topic: str = ""
    evidence: List[str] = Field(default_factory=list)
    debate_id: Optional[str] = None
    status: str = "unverified"  # unverified | disputed | verified | rejected
    created_at: float = Field(default_factory=time.time)


class ClaimLedger:
    def __init__(self):
        self._claims: Dict[str, Claim] = {}
        self._lock = asyncio.Lock()

    async def add_claim(
        self,
        text: str,
        confidence: int = 50,
        source_agent: str = "",
        topic: str = "",
        evidence: Optional[List[str]] = None,
    ) -> Claim:
        async with self._lock:
            claim = Claim(
                text=text,
                confidence=confidence,
                source_agent=source_agent,
                contributing_agents=[source_agent] if source_agent else [],
                topic=topic,
                evidence=evidence or [],
            )
            self._claims[claim.id] = claim
            return claim

    async def update_confidence(self, claim_id: str, new_confidence: int, agent_id: Optional[str] = None):
        async with self._lock:
            claim = self._claims.get(claim_id)
            if claim:
                claim.confidence = max(0, min(100, new_confidence))
                if agent_id and agent_id not in claim.contributing_agents:
                    claim.contributing_agents.append(agent_id)

    async def set_status(self, claim_id: str, status: str, debate_id: Optional[str] = None):
        async with self._lock:
            claim = self._claims.get(claim_id)
            if claim:
                claim.status = status
                if debate_id:
                    claim.debate_id = debate_id

    async def get_sorted(self) -> List[Claim]:
        async with self._lock:
            return sorted(self._claims.values(), key=lambda c: c.confidence, reverse=True)

    async def get_all(self) -> List[Claim]:
        async with self._lock:
            return list(self._claims.values())

    async def get_by_id(self, claim_id: str) -> Optional[Claim]:
        async with self._lock:
            return self._claims.get(claim_id)

    async def clear(self):
        async with self._lock:
            self._claims.clear()
